packages/superset-ui-core/src/callApi/parseResponse.js

```diff
@@ -1,26 +1,29 @@
-export default function parseResponse(apiPromise) {
-  return apiPromise.then(apiResponse =>
+const PARSERS = {
+  json: response =>
     // first try to parse as json, and fall back to text (e.g., in the case of HTML stacktrace)
     // cannot fall back to .text() without cloning the response because body is single-use
-    apiResponse
+    response
       .clone()
       .json()
-      .catch(() => /* jsonParseError */ apiResponse.text().then(textPayload => ({ textPayload })))
-      .then(maybeJson => ({
-        json: maybeJson.textPayload ? undefined : maybeJson,
-        response: apiResponse,
-        text: maybeJson.textPayload,
-      }))
-      .then(({ response, json, text }) => {
-        if (!response.ok) {
-          return Promise.reject({
-            error: response.error || (json && json.error) || text || 'An error occurred',
-            status: response.status,
-            statusText: response.statusText,
-          });
-        }
+      .then(json => ({ json, response }))
+      .catch(() => /* jsonParseError */ response.text().then(text => ({ response, text }))),
 
-        return typeof text === 'undefined' ? { json, response } : { response, text };
-      }),
-  );
+  text: response => response.text().then(text => ({ response, text })),
+};
+
+export default function parseResponse(apiPromise, parseMethod = 'json') {
+  const responseParser = PARSERS[parseMethod] || PARSERS.json;
```

Comment:
Can developer pass `parseMethod` as a custom function in addition to string of predefined parser name?

Reply:
I thought about that but am not sure a free-form parser makes sense. There are only a few available https://developer.mozilla.org/en-US/docs/Web/API/Response#Methods, of which we are exposing `text` and `json`. 

It doesn't seem worthwhile to copy / mirror the full response API to me (would begin questioning the need for this lib). One option could be to allow `null` or `'none'` as a `parseMethod` to enable returning the response unmodified and un-read. Beyond that, any custom parsing I think should come from a chained `.then(({ json }) => myCustomParser(json))`